libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+    }
+
+  squashed_ref_actions = g_ptr_array_new_full (g_hash_table_size (hash_table),
+                                               (GDestroyNotify) flatpak_remote_ref_action_unref);
+  g_hash_table_iter_init (&hash_iter, hash_table);
+
+  while (g_hash_table_iter_next (&hash_iter, &key, &value))
+    g_ptr_array_add (squashed_ref_actions,
+                     flatpak_remote_ref_action_ref (value));
+
+  g_ptr_array_sort (squashed_ref_actions, sort_flatpak_remote_ref_actions);
+
+  return squashed_ref_actions;
+}
+
+/* Given a hash table of filenames to FlatpakRemoteRefActionsFile, hoist
+ * the underlying pointer array of remote ref actions and make that the value
+ * of the new hash table.
+ *
+ * This will make the hash table suitable for passing to
+ * eos_updater_util_squash_remote_ref_actions */
+GHashTable *
+eos_updater_util_hoist_flatpak_remote_ref_actions (GHashTable *ref_actions_file_table)
+{
+  g_autoptr(GHashTable) hoisted_ref_actions_table = g_hash_table_new_full (g_str_hash,
+                                                                           g_str_equal,
+                                                                           g_free,
+                                                                           (GDestroyNotify) g_ptr_array_unref);
+  gpointer key, value;
+  GHashTableIter iter;
+
+  g_hash_table_iter_init (&iter, ref_actions_file_table);
+
+  while (g_hash_table_iter_next (&iter, &key, &value))
+    {
+      FlatpakRemoteRefActionsFile *ref_actions_file = value;
+
+      g_hash_table_insert (hoisted_ref_actions_table,
+                           g_strdup (key),
+                           g_ptr_array_ref (ref_actions_file->remote_ref_actions));
+    }
+
+  return g_steal_pointer (&hoisted_ref_actions_table);
+}
+
+/* Examine each of the remote ref action lists in ref_action_table
+ * and squash them down into a list where only one action is applied for
+ * each flatpak ref (the latest one) */
+GHashTable *
+eos_updater_util_squash_remote_ref_actions (GHashTable *ref_actions_table)
+{
+  gpointer key;
+  gpointer value;
+  GHashTableIter hash_iter;
+  GHashTable *squashed_ref_actions_table = g_hash_table_new_full (g_str_hash,
+                                                                  g_str_equal,
+                                                                  g_free,
+                                                                  (GDestroyNotify) g_ptr_array_unref);
+
+  g_hash_table_iter_init (&hash_iter, ref_actions_table);
+
+  while (g_hash_table_iter_next (&hash_iter, &key, &value))
+    {
+      GPtrArray *unsquashed_remote_ref_actions = value;
+      GPtrArray *squashed_remote_ref_actions = squash_ref_actions_ptr_array (unsquashed_remote_ref_actions);
+
+      g_hash_table_insert (squashed_ref_actions_table,
+                           g_strdup (key),
+                           squashed_remote_ref_actions);
+    }
```

Comment:
We need to squash the refs once, across all the serial number domains, not once per domain.

Reply:
Hmm, my understanding of the design doc was that we wanted to do this:

 - read in all serial domains
 - squash each one
 - concatenate all the squashed lists in lexicographical order
 - squash again

Based on the "Order of applying entries from autoinstall lists" section (see "Worked Example")

Reply:
I don’t think the first compression step (of individual files) is necessary. Are there any situations in which the final list of actions would differ depending on whether the first compression step was done?

I removed the first compression step from §(Worked example) mid-last-week. Sorry I didn’t point it out at the time. (You should probably re-check the worked example matches your understanding now.) (A diff view in Google Docs and an easy way to see ‘changes since I last viewed this document’ would be useful.)

Reply:
Off the top of my head, I don't *think* there would be any difference. @cosimoc  was the one who suggested the approach of compression twice, perhaps he can lend some insight?

Reply:
Ah damn, okay, we do need to do the initial squash.

It actually isn't so important for the step where the flatpaks are pulled, but it does matter for the step where the flatpaks are installed by `eos-updater-flatpak-installer`, since in that program, we do not assemble all actions into a single pointer array. This is because whilst we are running the installations, we need to keep track to see if any failed in a domain and if so update the counter accordingly. Therefore, the squash step needs to happen on each domain but not on the whole lot.

Reply:
Hmm. I’d prefer to apply the squash to the whole lot, since that could eliminate some actions which would cause apps to be downloaded unnecessarily (using the user’s bandwidth).

Three possible solutions to this:
 - Don’t update the stored serial number for that domain. This has the downsides we discussed before on the design document (§(Downloading and deploying flatpaks on OS upgrade)).
 - Re-parse the JSON files on failure to find the ID before the one which failed.
 - Store the ID from the previous action in a new field in our `struct` for storing a parsed action, and set the stored serial number to that on failure. (Better than option 2 since it doesn’t require re-parsing.)

I think I prefer the third option, and it should be fairly simple to implement, I think?